### ISSUE TITLE
Error in emcc.py with "native-wasm" setting

### ISSUE THREAD

**Issue by Akdez** [emscripten-core/emscripten#5743]
Hello.

I've met an error in emcc.py in line 887:

`setattr(shared.Settings, key, eval(value))`

Python log:
```
Traceback (most recent call last):
  File "E:\Projects\oryol\fips-sdks\win\emsdk-portable\emscripten\incoming\\em++", line 16, in <module>
    emcc.run()
  File "E:\Projects\oryol\fips-sdks\win\emsdk-portable\emscripten\incoming\emcc.py", line 892, in run
    setattr(shared.Settings, key, eval(value))
  File "<string>", line 1
    native-wasm'
               ^
SyntaxError: EOL while scanning string literal
```

Error occurred when values was:
`key = 'BINARYEN_METHOD`
`value = native-wasm'`

Function "eval" obviously failing to proceed with "native-wasm'" string.

**Comment by kripken** [comment]
What is the full emcc command you issued? This might be due to how you escaped the string. You may need to do something like  `-s 'BINARYEN_METHOD='native-wasm"'`.

**Comment by Akdez** [comment]
Here is the full command line:

```
cmd.exe /C "cd . && E:\Projects\oryol\fips-sdks\win\emsdk-portable\emscripten\incoming\em++.bat  --em-config E:/Projects/oryol/fips-sdks/win/emsdk-portable/.emscripten --cache E:/Projects/oryol/fips-sdks/win/emsdk-portable/.emscripten_cache  -fno-exceptions -fno-rtti -std=c++11 -stdlib=libc++ -fstrict-aliasing -Wall -Wno-multichar -Wextra -Wno-unused-parameter -Wno-unknown-pragmas -Wno-ignored-qualifiers -Wno-long-long -Wno-overloaded-virtual -Wno-deprecated-writable-strings -Wno-unused-volatile-lvalue -Wno-inconsistent-missing-override -Wno-warn-absolute-paths -Wno-expansion-to-defined -O3 -DNDEBUG  --em-config E:/Projects/oryol/fips-sdks/win/emsdk-portable/.emscripten --cache E:/Projects/oryol/fips-sdks/win/emsdk-portable/.emscripten_cache  --memory-init-file 1 -s TOTAL_MEMORY=134217728 -s ERROR_ON_UNDEFINED_SYMBOLS=1 -s NO_EXIT_RUNTIME=1 -s USE_WEBGL2=1 -s FULL_ES3=1 -s NO_FILESYSTEM=1 -s DISABLE_EXCEPTION_CATCHING=1 -s WASM=1 -s 'BINARYEN_METHOD="native-wasm"'  --shell-file E:/Projects/oryol/fips-sdks/win/emsdk-portable/emscripten/incoming/src/shell_minimal.html -O3  --llvm-lto 1 -s OUTLINING_LIMIT=60000 --closure 0 -s ASSERTIONS=0 code\Samples\IOQueueSample\CMakeFiles\IOQueueSample.dir\IOQueueSample.cc.o  -o E:\Projects\oryol\fips-deploy\oryol\webgl2-wasm-ninja-release\IOQueueSample.html  code\Modules\IO\libIO.a code\Modules\HttpFS\libHttpFS.a code\Modules\IO\libIO.a code\Modules\Core\libCore.a && cd ."
```

There is that option exactly:
`-s 'BINARYEN_METHOD="native-wasm"'`

P.S.
Sorry, it was not exactly written by me, it's from build tool called "fips" - http://floooh.github.io/fips/index.html

I were building some samples of one project that uses fips and emscripten when I've met this error so I decided to let you know about it.

**Comment by kripken** [comment]
The quotes in that line look ok to me, and they work on my machine (linux/dash). Perhaps escaping works differently on windows?

**Comment by Akdez** [comment]
I'm not sure if it's about escaping or something like that. I've also tried using it without quotes
`-s BINARYEN_METHOD="native-wasm"`

And there was another error:
```
Traceback (most recent call last):
  File "E:\Projects\oryol\fips-sdks\win\emsdk-portable\emscripten\incoming\\em++", line 16, in <module>
    emcc.run()
  File "E:\Projects\oryol\fips-sdks\win\emsdk-portable\emscripten\incoming\emcc.py", line 892, in run
    setattr(shared.Settings, key, eval(value)) 
  File "<string>", line 1, in <module>
NameError: name 'native' is not defined
```

As you can see the function eval(value) trying to evaluate "native" minus "wasm".
But it works fine without eval() function:
setattr(shared.Settings, key, eval(value) if value != 'native-wasm' else value)

**Comment by floooh** [comment]
Hi @kripken, I think you can close this issue, it's something I need to take care of in the fips cmake files :)

The problem seem to be those strange ' quotes: ```-s 'BINARYEN_METHOD="native-wasm"'```, and this seems to happen only on Windows. I vaguely remember having seen a similar problem in the past.

TL;DR: I will take over, it's not an emscripten issue ;)

**Comment by floooh** [comment]
...hm or maybe it's not so simple... the reason those two different quote types are necessary is described here:

https://github.com/kripken/emscripten/wiki/WebAssembly#binaryen-methods

Removing the outer quotes indeed doesn't work (in an OSX shell at least).

I seem to remember that the BINARYEN_METHOD flag was different from the other -s flags because of they way those args are handed over to the binaryen tool by emcc... I'll investigate a bit more...

**Comment by floooh** [comment]
...ok I should pay more attention to the docs :)

Looks like ```-s "BINARYEN_METHOD='native-wasm'"``` works on Windows, the problem happens with ```-s 'BINARYEN_METHOD="native-wasm"'```.

I just messed up the quote types in fips...

**Comment by kripken** [comment]
Ok, thanks for the info. Weird that windows does that...

So the only cross-platform way to write this is `-s "BINARYEN_METHOD='native-wasm'"`? We should look for places in the docs where we get that wrong.

**Comment by floooh** [comment]
Yes, looks like this is the only correct way. I think what happens is this:

- somewhere the double-quotes seem to get lost (I guess in cmd.exe, since that's the only Windows-specific part), but the single-quotes are kept in
- the value part in the -s key/value pair runs through python "eval()" (I guess to get strings as python strings and number strings as python numbers)
- for most simple values this isn't a problem, since python seems to do the right thing also without the single-quotes
- a string like 'native-wasm' however looks like a math expression to python now when running through eval

...IMHO (if this part makes trouble in the future again) a more robust way would be to get rid of the eval(), and do some sort of parsing on the value item (for instance check if it is a valid number, otherwise treat as string, I'm not sure if a value item can also be evaluated into a python list currently though)
